Missing hooks on a resource pass validation

`Resource.InternalValidate` does not seem to catch if a `*schema.Resource` doesn't have CRUD hooks defined. If used, it gives the following panic:

```
=== RUN   TestAccTritonVLAN_basic
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x129db9]

goroutine 1059 [running]:
panic(0x6cec40, 0xc82000e0f0)
        /usr/local/Cellar/go/1.6/libexec/src/runtime/panic.go:464 +0x3e6
github.com/hashicorp/terraform/helper/schema.(*Resource).Apply(0xc82010a2c0, 0xc820506c30, 0xc82017f060, 0x7aefc0, 0xc82017faf0, 0x119f01, 0x0, 0x0)
        /Users/brianhicks/code/go/src/github.com/hashicorp/terraform/helper/schema/resource.go:146 +0x299
github.com/hashicorp/terraform/helper/schema.(*Provider).Apply(0xc82010e180, 0xc820210000, 0xc820506c30, 0xc82017f060, 0x1, 0x0, 0x0)
        /Users/brianhicks/code/go/src/github.com/hashicorp/terraform/helper/schema/provider.go:162 +0x1ed
github.com/hashicorp/terraform/terraform.(*EvalApply).Eval(0xc820514300, 0x1906068, 0xc82034c1a0, 0x0, 0x0, 0x0, 0x0)
        /Users/brianhicks/code/go/src/github.com/hashicorp/terraform/terraform/eval_apply.go:67 +0x6e6
github.com/hashicorp/terraform/terraform.EvalRaw(0x1984320, 0xc820514300, 0x1906068, 0xc82034c1a0, 0x0, 0x0, 0x0, 0x0)
        /Users/brianhicks/code/go/src/github.com/hashicorp/terraform/terraform/eval.go:53 +0x25b
github.com/hashicorp/terraform/terraform.(*EvalSequence).Eval(0xc8204d0600, 0x1906068, 0xc82034c1a0, 0x0, 0x0, 0x0, 0x0)
        /Users/brianhicks/code/go/src/github.com/hashicorp/terraform/terraform/eval_sequence.go:10 +0xea
github.com/hashicorp/terraform/terraform.EvalRaw(0x1940118, 0xc8204d0600, 0x1906068, 0xc82034c1a0, 0x0, 0x0, 0x0, 0x0)
        /Users/brianhicks/code/go/src/github.com/hashicorp/terraform/terraform/eval.go:53 +0x25b
github.com/hashicorp/terraform/terraform.(*EvalOpFilter).Eval(0xc820506390, 0x1906068, 0xc82034c1a0, 0x0, 0x0, 0x0, 0x0)
        /Users/brianhicks/code/go/src/github.com/hashicorp/terraform/terraform/eval_filter_operation.go:37 +0x67
github.com/hashicorp/terraform/terraform.EvalRaw(0x1940140, 0xc820506390, 0x1906068, 0xc82034c1a0, 0x0, 0x0, 0x0, 0x0)
        /Users/brianhicks/code/go/src/github.com/hashicorp/terraform/terraform/eval.go:53 +0x25b
github.com/hashicorp/terraform/terraform.(*EvalSequence).Eval(0xc8204844a0, 0x1906068, 0xc82034c1a0, 0x0, 0x0, 0x0, 0x0)
        /Users/brianhicks/code/go/src/github.com/hashicorp/terraform/terraform/eval_sequence.go:10 +0xea
github.com/hashicorp/terraform/terraform.EvalRaw(0x1940118, 0xc8204844a0, 0x1906068, 0xc82034c1a0, 0x0, 0x0, 0x0, 0x0)
        /Users/brianhicks/code/go/src/github.com/hashicorp/terraform/terraform/eval.go:53 +0x25b
github.com/hashicorp/terraform/terraform.Eval(0x1940118, 0xc8204844a0, 0x1906068, 0xc82034c1a0, 0x0, 0x0, 0x0, 0x0)
        /Users/brianhicks/code/go/src/github.com/hashicorp/terraform/terraform/eval.go:34 +0x6a
github.com/hashicorp/terraform/terraform.(*Graph).walk.func1(0x778520, 0xc820409830, 0x0, 0x0)
        /Users/brianhicks/code/go/src/github.com/hashicorp/terraform/terraform/graph.go:198 +0x9a9
github.com/hashicorp/terraform/dag.(*AcyclicGraph).Walk.func3(0xc820258a10, 0xc8204e1400, 0xc820258028, 0xc820409950, 0xc820258a30, 0x778520, 0xc820409830, 0xc8204df980, 0xc8204df9e0)
        /Users/brianhicks/code/go/src/github.com/hashicorp/terraform/dag/dag.go:238 +0xeb
created by github.com/hashicorp/terraform/dag.(*AcyclicGraph).Walk
        /Users/brianhicks/code/go/src/github.com/hashicorp/terraform/dag/dag.go:247 +0x715
exit status 2
```

When I add those hooks, this panic goes away. I'm working on a provider based on commit 1c399d119a796e313f9f20201a156da866ee7ace

(reported initially to @jen20, who suggested I open an issue)

--- a/helper/schema/resource.go
+++ b/helper/schema/resource.go
@@ -317,6 +317,14 @@ func (r *Resource) InternalValidate(topSchemaMap schemaMap, writable bool) error
 
 		tsm = schemaMap(r.Schema)
 
+		// Destroy, and Read are required
+		if r.Read == nil {
+			return fmt.Errorf("Read must be implemented")
+		}
+		if r.Delete == nil {
+			return fmt.Errorf("Delete must be implemented")
+		}
+
 		// If we have an importer, we need to verify the importer.
 		if r.Importer != nil {
 			if err := r.Importer.InternalValidate(); err != nil {

--- a/helper/schema/resource_test.go
+++ b/helper/schema/resource_test.go
@@ -480,13 +480,49 @@ func TestResourceInternalValidate(t *testing.T) {
 			false,
 			true,
 		},
+
+		// writable must have Read
+		{
+			&Resource{
+				Create: func(d *ResourceData, meta interface{}) error { return nil },
+				Update: func(d *ResourceData, meta interface{}) error { return nil },
+				Delete: func(d *ResourceData, meta interface{}) error { return nil },
+				Schema: map[string]*Schema{
+					"goo": &Schema{
+						Type:     TypeInt,
+						Optional: true,
+					},
+				},
+			},
+			true,
+			true,
+		},
+
+		// writable must have Delete
+		{
+			&Resource{
+				Create: func(d *ResourceData, meta interface{}) error { return nil },
+				Read:   func(d *ResourceData, meta interface{}) error { return nil },
+				Update: func(d *ResourceData, meta interface{}) error { return nil },
+				Schema: map[string]*Schema{
+					"goo": &Schema{
+						Type:     TypeInt,
+						Optional: true,
+					},
+				},
+			},
+			true,
+			true,
+		},
 	}
 
 	for i, tc := range cases {
-		err := tc.In.InternalValidate(schemaMap{}, tc.Writable)
-		if err != nil != tc.Err {
-			t.Fatalf("%d: bad: %s", i, err)
-		}
+		t.Run(fmt.Sprintf("#%d", i), func(t *testing.T) {
+			err := tc.In.InternalValidate(schemaMap{}, tc.Writable)
+			if err != nil != tc.Err {
+				t.Fatalf("%d: bad: %s", i, err)
+			}
+		})
 	}
 }
 

